csharp/src/Apache.Arrow/Types/TimestampType.cs

```diff
@@ -27,6 +29,15 @@ public sealed class TimestampType : FixedWidthType
         public TimeUnit Unit { get; }
         public string Timezone { get; }
 
+        public TimeZoneInfo GetTimeZoneInfo()
+        {
+            return !string.IsNullOrEmpty(Timezone)
+                ? TimeZoneInfo.FindSystemTimeZoneById(Timezone)
```

Comment:
One issue here is that `TimeZoneInfo.FindSystemTimeZoneById` behaves differently on Unix vs. Windows. See [https://github.com/dotnet/corefx/issues/2538](https://github.com/dotnet/corefx/issues/2538).

Maybe we just expose `string Timezone` and callers can do what they want with it (like use [NodaTime](https://nodatime.org/) or something else).

Reply:
The C++ docs say:

> If supplied, the timezone string should take either the form (i) “Area/Location”, with values drawn from the names in the IANA Time Zone Database (such as “Europe/Zurich”); or (ii) “(+|-)HH:MM” indicating an absolute offset from GMT (such as “-08:00”). To indicate a native UTC timestamp, one of the strings “UTC”, “Etc/UTC” or “+00:00” should be used.

Do we support the `(+|-)HH:MM` notation?

See [https://arrow.apache.org/docs/cpp/api/datatype.html#_CPPv4N5arrow13TimestampTypeE](https://arrow.apache.org/docs/cpp/api/datatype.html#_CPPv4N5arrow13TimestampTypeE)

---
In reply to: [355715543](https://github.com/apache/arrow/pull/5965#discussion_r355715543) [](ancestors = 355715543)

Reply:
Exposing just the Timezone string would be okay, although I am curious about how much input validation we should do. Do you think it would make sense to implement our own IANA Time Zone Database provider? (or find a suitable dependency)

Considering that the CoreFX implementation of TimeZoneInfo is not necessarily consistent with the IANA format (and may not ever be, for compatibility reasons).

Reply:
IMO - this is where NodaTime fits it. Callers can use that to convert IANA time zone IDs (like `Europe/Zurich`) to offsets.

However, I don't think the Arrow library should depend on NodaTime, since I assume a considerable amount of users aren't interested in TimeStamps. Or if they are, they always use offsets (`-08:00`) and won't need to look up the IDs.

That's why I think the Arrow library provides the data, ex. just `string Timezone`, and callers can optionally opt into NodaTime (or something else) to do the time zone lookup.

Reply:
Do you have any thoughts on handling **TimestampArray#GetTimestamp**? At the moment the implementation attempts to parse the timezone information and then applies the relevant offset on calls to **GetTimestamp**, since the underlying timestamp is normalized to UTC.

I see a few options:

- Constructor argument for the desired time zone
- Argument to **GetTimestamp** specifying the desired time zone (could override the constructor argument)
- Provide an interface that the array implementation can query against to resolve an IANA timezone string to an appropriate **TimeZoneInfo** structure, and in the future provide optional auxiliary packages that implement the interface (e.g. **Apache.Arrow.NodaTime**).
- Do nothing and always return UTC **DateTimeOffset**, and require caller to convert to the expected time zone (e.g. **ToLocalTime**, **OffsetTo**)

Reply:
@eerhardt I have made the following changes:

1. When constructing a timestamp array, the **TimeZoneInfo** is converted to an ISO8601 offset string (for convenience)
2. **TimestampArray#GetTimestamp** no longer attempts to apply an offset (the result is always UTC normalized)
3. **TimestampArray#GetTimestampUnchecked** provides a path that does not perform a null check

Reply:
This all sounds good to me.

As for the name of `GetTimestampUnchecked`, I'd be fine with what you have. The only alternative I can think of would be `GetTimestampSkipNullCheck`. I'm not confident that name is better than what you have.